Algorithm Testing (#1033)
opendcs/opendcs: docs/source/dev-docs.rst

|                                            |case the sets will be merged    |
+--------------------------------------------+--------------------------------+

Adding Subtests
---------------

This module provides functionality for decoding data and implementing algorithms.

Decodes Functions
~~~~~~~~~~~~~~~~~

> - **Purpose**: Functions in this section are responsible for decoding input data into a usable format.
> - **Usage**: These functions are typically used to parse encoded data or transform it into a structured representation.

Reviewer: ```suggestion
```

Adding Tests for Decodes Functions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
To add new tests for Decodes Functions, developers need to create four files in the `/git/opendcs/java/opendcs/src/test/resources/decodes/db` directory:

1. **`.assertions` file**:  

>     - Purpose: Defines the expected output or results for the test.  
>     - Usage: Specify the expected values for the decoded data to validate the function's correctness.  

Reviewer: ```suggestion
    - Purpose: Defines the expected output for the test to validate the Decoding.
```

         2,2014-03-01T14:00:00Z,15.2,  0.0, Expected value not parsed (sensor 2)
         3,2014-03-01T12:00:00Z,98.1,  0.0, Expected value not parsed (sensor 3)
         3,2014-03-01T13:00:00Z,98.1,  0.0, Expected value not parsed (sensor 3)
         3,2014-03-01T14:00:00Z,98.2,  0.0, Expected value not parsed (sensor 3)
         4,2014-03-01T12:00:00Z,8252,  0.0, Expected value not parsed (sensor 4)
         4,2014-03-01T13:00:00Z,8252,  0.0, Expected value not parsed (sensor 4)
         4,2014-03-01T14:00:00Z,8252,  0.0, Expected value not parsed (sensor 4)
         5,2014-03-01T12:00:00Z,0.0,  0.0, Expected value not parsed (sensor 5)
         5,2014-03-01T13:00:00Z,0.0,  0.0, Expected value not parsed (sensor 5)
         5,2014-03-01T14:00:00Z,0.0,  0.0, Expected value not parsed (sensor 5)
         6,2014-03-01T12:00:00Z,0.0,  0.0, Expected value not parsed (sensor 6)
         6,2014-03-01T13:00:00Z,0.0,  0.0, Expected value not parsed (sensor 6)
         6,2014-03-01T14:00:00Z,0.0,  0.0, Expected value not parsed (sensor 6)

2. **`.decodescript` file**:  
    - Purpose: Contains the Decodes script that defines how the input data should be processed.  

>     - Usage: Write the script to test the specific functionality of the Decodes Function.  

Reviewer: ```suggestion
```

    - Example:  
      .. code-block:: text

         csv: 3(/, F(D,A,10,4), x, F(T,A,8), csv(1, 2, 4, 5, 6, 3))

3. **`.input` file**:  
    - Purpose: Provides the raw input data to be decoded.  

>     - Usage: Include the encoded data that the Decodes Function will process.  

Reviewer: ```suggestion
```

    - Example:  
      .. code-block:: text

         # Ignored header line
         03/01/2014 12:00:00 23.95, 17.2, 8252, 0, 0, 98.1
         03/01/2014 13:00:00 23.96, 16.9, 8252, 0, 0, 98.1
         03/01/2014 14:00:00 23.97, 15.2, 8252, 0, 0, 98.2

4. **`.sensors` file**:  
    - Purpose: Describes the sensors and their configurations used in the decoding process.  

>     - Usage: Define the sensor metadata required for the test.  

Reviewer: ```suggestion
```

    - Example:  
      .. code-block:: csv

         #sensor number, sensor name, units, description
         1, Stage, ft, none
         2, Humidity, %, none
         3, Temp, degF, none
         4, Storage, acft, none
         5, Precip, in, none
         6, Zero, raw, none

> By adding these files, developers can create comprehensive tests to ensure the correctness and reliability of new or modified Decodes Functions.

Reviewer: ```suggestion
By adding these files, developers can create tests to ensure the correctness and reliability of the Decodes language including new or modified Decodes Functions.
```

Algorithms
~~~~~~~~~~

> - **Purpose**: This section contains algorithmic implementations for solving specific problems or performing computations.

Reviewer: ```suggestion
- **Purpose**: This section describes how to test algorithmic implementations for solving specific problems or performing computations.
```

> - **Usage**: These algorithms are designed to be reusable and efficient for various applications.

Reviewer: ```suggestion
```

             2024/10/05-04:00:00,25.9200000000,0
             2024/10/05-05:00:00,63.0000000001,0
             2024/10/05-06:00:00,59.4000000000,0
             2024/10/05-07:00:00,50.0400000001,0
             2024/10/05-08:00:00,40.6800000000,0
             2024/10/05-09:00:00,59.4000000000,0
             2024/10/05-10:00:00,38.8800000000,0
             2024/10/05-11:00:00,20.5200000000,0
             2024/10/05-12:00:00,25.9200000000,0
             2024/10/05-13:00:00,18.3600000000,0
             2024/10/05-14:00:00,16.5600000001,0

          You can create this file using the `outputts` command.

        - `output` directory: Contains `.tsimport` files defining the output time series generated by the computation.  
        - `expectedOutputs` directory: Contains `.tsimport` files defining the expected output time series for validation.

3. **Computation Configuration**:  
    - File: `comp.xml`  
    - Purpose: This file defines the setup for the computation and the tests that need to be run. It specifies the algorithm configuration and links the input, output, and expected output time series.  
    - Note: You can create the `comp.xml` file using the `compexport` command.

> By organizing the test resources in this structure, developers can ensure that the algorithm tests are comprehensive and easy to manage.

Reviewer: ```suggestion
By organizing the algorithm test resources in this structure, developers can easily manage algorithm tests.
```
